### PR TITLE
LSP xml server and client

### PR DESCRIPTION
**Pull request by sapirpol**

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (bd82) [inline .vscode/launch.json]: launch configuration does not belong in top of repo

File: packages/vscode-xml-tools-client/.gitignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+out
+node_modules
```
Comment (bd82) [inline packages/vscode-xml-tools-client/.gitignore]: node_modules is defined in higher level ignore

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  // Use IntelliSense to learn about possible attributes.
+  // Hover to view descriptions of existing attributes.
+  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+  "version": "0.2.0",
+  "configurations": [
+    {
+      "type": "node",
+      "request": "launch",
+      "name": "Launch Program",
+      "program": "${workspaceFolder}/packages/vscode-xml-tools-client/src/extension.ts",
```
Comment (bd82) [inline .vscode/launch.json]: i think we need a shorter name for the extension

File: packages/vscode-xml-tools-client/.vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "editor.insertSpaces": false,
+  "tslint.enable": true,
+  "typescript.tsc.autoDetect": "off",
+  "typescript.preferences.quoteStyle": "single"
```
Comment (bd82) [inline packages/vscode-xml-tools-client/.vscode/settings.json]: formatting options should be handled by prettier

File: packages/vscode-xml-tools-client/.vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "editor.insertSpaces": false,
```
Comment (bd82) [inline packages/vscode-xml-tools-client/.vscode/settings.json]: is `insertSpaces` a formatting option? if so let prettier handle it

File: packages/vscode-xml-tools-client/.vscode/tasks.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (bd82) [inline packages/vscode-xml-tools-client/.vscode/tasks.json]: is this needed? it is just a duplicate/ref-to of the package.json scripts

File: packages/vscode-xml-tools-client/.vscodeignore [hunk 7]
```diff
@@ -0,0 +1,14 @@
+.vscode/**
```
Comment (bd82) [inline packages/vscode-xml-tools-client/.vscodeignore]: why do we need this file?

File: packages/vscode-xml-tools-client/package.json [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "vscode-xml-tools-client",
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: This is the name of the extension, it should be shorter for consumption in VSCode Marketplace.
- which means name here may not be equal to the folder name

File: packages/vscode-xml-tools-client/package.json [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  "description": "VSCode client part of an xml language server",
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: description should be written as facing consumer from VSCode marketplace.
basically not in technical terms

File: packages/vscode-xml-tools-client/package.json [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  "author": "SAP SE",
+  "version": "0.0.1",
+  "publisher": "vscode",
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: the publisher is not vscode....

File: packages/vscode-xml-tools-client/package.json [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  "main": "./out/extension",
+  "scripts": {
+    "vscode:prepublish": "npm run update-vscode && cd .. && npm run compile",
+    "postinstall": "vscode-install",
+    "compile": "tsc -b",
+    "watch": "tsc -b -w",
+    "build": "vsce package"
+  },
+  "dependencies": {
+    "vscode-languageclient": "^5.2.1"
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: use fixed version for dep, we have dependentBot to update versions

File: packages/vscode-xml-tools-client/package.json [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  "devDependencies": {
+    "@types/node": "^12.11.7",
+    "vscode": "^1.1.35",
+    "typescript": "^3.6.4"
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: use latest versions of devDeps

File: packages/vscode-xml-tools-client/package.json [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    "postinstall": "vscode-install",
+    "compile": "tsc -b",
+    "watch": "tsc -b -w",
+    "build": "vsce package"
+  },
+  "dependencies": {
+    "vscode-languageclient": "^5.2.1"
+  },
+  "devDependencies": {
+    "@types/node": "^12.11.7",
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: we are targeting node 10, so do not use node 12 types

File: packages/vscode-xml-tools-client/tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es6",
```
Comment (bd82) [inline packages/vscode-xml-tools-client/tsconfig.json]: target should be es7 I think

File: packages/vscode-xml-tools-client/tslint.json [hunk 15]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (bd82) [inline packages/vscode-xml-tools-client/tslint.json]: tslint is deprecated.
consider if you want to spend time on this or just use JavaScript instead.

File: packages/xml-lsp-server/package.json [hunk 16]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "@xml-tools/xml-lsp-server",
+  "description": "Generic xml language server",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: Generic is probably a technical term, does not fit in description

File: packages/xml-lsp-server/package.json [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  "version": "1.0.0",
+  "author": "SAPSE",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: this is not how you wrote SAP-SE earlier

Comment (bd82) [reply packages/xml-lsp-server/package.json]: maybe remove this property? how did we do in the other packages?

File: packages/xml-lsp-server/package.json [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+  "license": "Apache-2.0",
+  "engines": {
+    "node": "*"
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: we are targeting node 10 not *

File: packages/xml-lsp-server/package.json [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/SAP/xml-tools/"
+  },
+  "dependencies": {
+    "vscode-languageserver": "^5.2.1",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: fixed versions please

File: packages/xml-lsp-server/package.json [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    "@xml-tools/parser": "^1.0.1"
+  },
+  "devDependencies": {
+    "@types/mocha": "^5.2.7",
+    "@types/node": "^12.11.7",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: node types version 10 please.

File: packages/xml-lsp-server/package.json [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    "typescript": "^3.6.4"
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: this is not latest typescript version

File: packages/xml-lsp-server/tsconfig.json [hunk 22]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es6",
```
Comment (bd82) [inline packages/xml-lsp-server/tsconfig.json]: es7

File: packages/xml-lsp-server/src/server.ts [hunk 23]
```diff
@@ -0,0 +1,132 @@
+import {
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: i will go over this later

File: packages/vscode-xml-tools-client/tslint.json (hunk 15 above)
Comment (bd82) [reply packages/vscode-xml-tools-client/tslint.json]: note your client has tslint but your server does not

File: packages/vscode-xml-tools-client/src/extension.ts [hunk 24]
```diff
@@ -0,0 +1,70 @@
+import * as path from "path";
```
Comment (bd82) [inline packages/vscode-xml-tools-client/src/extension.ts]: i will go over this later

File: packages/vscode-xml-tools-client/package.json [hunk 25]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "vscode-xml-tools-client",
+  "description": "VSCode client part of an xml language server",
+  "author": "SAP SE",
```
Comment (bd82) [inline packages/vscode-xml-tools-client/package.json]: maybe remove author field

File: packages/xml-lsp-server/tsconfig.json (hunk 22 above)
Comment (rayakoren) [reply packages/xml-lsp-server/tsconfig.json]: Does not compile with es7 ...

File: packages/xml-lsp-server/.eslintrc.json [hunk 26]
```diff
@@ -8,7 +8,11 @@
     "plugin:prettier/recommended"
   ],
   "rules": {
-    "@typescript-eslint/no-use-before-define": ["error", { "functions": false }]
+    "@typescript-eslint/no-use-before-define": [
+      "error",
+      { "functions": false }
+    ],
+    "@typescript-eslint/no-explicit-any": ["off"]
```
Comment (bd82) [inline packages/xml-lsp-server/.eslintrc.json]: seems strange to disable this rule everywhere

File: packages/xml-lsp-server/nyc.config.js [hunk 27]
```diff
@@ -3,11 +3,12 @@ const nycConfigTypeScript = require("@istanbuljs/nyc-config-typescript");
...
   include: ["src/**/*.ts"],
+  exclude: ["src/server.ts"],
   reporter: ["text", "lcov"],
   "check-coverage": true,
   all: true,
   branches: 100,
   lines: 100,
-  functions: 100,
-  statements: 100
+  functions: 83,
```
Comment (bd82) [inline packages/xml-lsp-server/nyc.config.js]: This conflicts with the guidelines of 100%* coverage for this repo
- https://github.com/SAP/xml-tools/blob/master/CONTRIBUTING.md#test-coverage

We should exclude things that are not covered (assuming there is a good reason)

File: packages/xml-lsp-server/out/server.js [hunk 28]
```diff
@@ -0,0 +1,139 @@
+"use strict";
```
Comment (bd82) [inline packages/xml-lsp-server/out/server.js]: Generated source code should not be merged

File: packages/xml-lsp-server/out/server.js.map [hunk 29]
```diff
@@ -0,0 +1 @@
...,WAAW,EAAE,CAAC,CAAC;SAChE;IACH,CAAC;CAAA;AAED,MAAM,uBAAuB,GAAG,CAAC,QAAsB,EAAE,EAAE,CAAC,CAC1D,KAAmB,EACP,EAAE,CAAC,CAAC;IAChB,OAAO,EAAE,KAAK,CAAC,OAAO;IACtB,KAAK,EAAE,6BAAK,CAAC,MAAM,CACjB,QAAQ,CAAC,UAAU,CAAC,KAAK,CAAC,MAAM,CAAC,EACjC,QAAQ,CAAC,UAAU,CAAC,KAAK,CAAC,MAAM,GAAG,KAAK,CAAC,MAAM,CAAC,CACjD;IACD,QAAQ,EAAE,0CAAkB,CAAC,KAAK;IAClC,MAAM,EAAE,gBAAgB;CACzB,CAAC,CAAC;AAEH,MAAM,wBAAwB,GAAG,CAAC,QAAsB,EAAE,EAAE,CAAC,CAC3D,KAA4B,EAChB,EAAE,CAAC,CAAC;IAChB,OAAO,EAAE,KAAK,CAAC,OAAO;IACtB,KAAK,EAAE;QACL,KAAK,EAAE,QAAQ,CAAC,UAAU,CAAC,KAAK,CAAC,KAAK,CAAC,WAAW,CAAC;QACnD,GAAG,EAAE,QAAQ,CAAC,UAAU,CAAC,KAAK,CAAC,KAAK,CAAC,SAAU,CAAC;KACjD;IACD,QAAQ,EAAE,0CAAkB,CAAC,KAAK;IAClC,MAAM,EAAE,gBAAgB;CACzB,CAAC,CAAC;AAEH,SAAS,CAAC,SAAS,CAAC,CAAC,KAA8B,EAAE,EAAE;IACrD,IAAI,KAAK,CAAC,QAAQ,CAAC,UAAU,KAAK,KAAK,EAAE;QACvC,gBAAgB,CAAC,GAAG,CAAC,KAAK,CAAC,QAAQ,CAAC,GAAG,EAAE,IAAI,GAAG,EAAE,CAAC,CAAC;KACrD;AACH,CAAC,CAAC,CAAC;AAEH,SAAS,CAAC,UAAU,CAAC,CAAC,KAA8B,EAAE,EAAE;IACtD,IAAI,KAAK,CAAC,QAAQ,CAAC,UAAU,KAAK,KAAK,EAAE;QACvC,gBAAgB,CAAC,MAAM,CAAC,KAAK,CAAC,QAAQ,CAAC,GAAG,CAAC,CAAC;KAC7C;AACH,CAAC,CAAC,CAAC;AAEH,qDAAqD;AACrD,SAAS,CAAC,MAAM,CAAC,UAAU,CAAC,CAAC;AAE7B,2BAA2B;AAC3B,UAAU,CAAC,MAAM,EAAE,CAAC"}
```
Comment (bd82) [inline packages/xml-lsp-server/out/server.js.map]: source maps should not be merged

File: packages/xml-lsp-server/package.json [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+    "@istanbuljs/nyc-config-typescript": "1.0.1",
+    "@types/mocha": "7.0.1",
+    "@types/node": "10.17.15",
+    "@types/vscode": "1.32.0",
+    "@typescript-eslint/parser": "2.19.0",
+    "@typescript-eslint/eslint-plugin": "2.19.0",
+    "eslint": "6.8.0",
+    "eslint-config-prettier": "6.10.0",
+    "eslint-plugin-prettier": "3.1.2",
+    "mocha": "7.0.1",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: several of these devDepedencie are listed on the root level package.
Avoid duplicating these here.

File: packages/xml-lsp-server/package.json [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+    "sinon": "7.3.2",
+    "source-map-support": "0.5.16",
+    "ts-node": "8.6.2",
+    "typescript": "3.7.5",
+    "vscode": "^1.1.35",
+    "vscode-test": "1.3.0",
+    "vscode-languageclient": "5.2.1"
+  },
+  "scripts": {
+    "compile": "tsc -b",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: a `ci` script is needed to integrate into the mono-repo build

File: packages/xml-lsp-server/src/server.ts [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+let hasConfigurationCapability = false;
+let hasWorkspaceFolderCapability = false;
+
+connection.onInitialize((params: InitializeParams) => {
+  const capabilities = params.capabilities;
+
+  // Does the client support the `workspace/configuration` request?
+  // If not, we will fall back using global settings
+  hasConfigurationCapability = !!(
+    capabilities.workspace && !!capabilities.workspace.configuration
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: Would `capabilitties?.workspace?.cofiguration === true` suffice here?

File: packages/xml-lsp-server/src/server.ts [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+  );
+  hasWorkspaceFolderCapability = !!(
+    capabilities.workspace && !!capabilities.workspace.workspaceFolders
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: same as above, use `optional chaining`

File: packages/xml-lsp-server/src/server.ts [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  };
+});
+
+connection.onInitialized(() => {
+  if (hasConfigurationCapability) {
+    // Register for all configuration changes
+    connection.client.register(
+      DidChangeConfigurationNotification.type,
+      undefined
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: what is this `undefined` argument?

File: packages/xml-lsp-server/src/server.ts [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+    );
+  }
+  if (hasWorkspaceFolderCapability) {
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: if we are not doing anything with this event currently, do we need the event handler?

File: packages/xml-lsp-server/.eslintrc.json (hunk 26 above)
Comment (rayakoren) [reply packages/xml-lsp-server/.eslintrc.json]: The linter failed when I explicitly defined an 'any' type

File: packages/xml-lsp-server/src/server.ts [hunk 36]
```diff
@@ -0,0 +1,76 @@
...
+      } else if (_event.removed.length > 0) {
+        connection.console.log("Workspace folders were removed");
+      }
+    });
+  }
+});
+
+// The content of a text document has changed
+documents.onDidChangeContent((event: TextDocumentChangeEvent) => {
+  validateDocument(connection, event.document);
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: maybe make the test that the document is an XML doc here and not inside `validateDocument`

File: packages/xml-lsp-server/src/server.ts [hunk 37]
```diff
@@ -0,0 +1,76 @@
...
+});
+
+documents.onDidOpen((event: TextDocumentChangeEvent) => {
+  if (event.document.languageId === "xml") {
+    diagnosticsCache.set(event.document.uri, new Map());
+  }
+});
+
+documents.onDidClose((event: TextDocumentChangeEvent) => {
+  if (event.document.languageId === "xml") {
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: do we need to test the `languageId` everytime or could we assume the LSP Client or LSP library already does this filtering for us? basically what happens in other Language servers?

File: packages/xml-lsp-server/test/server.spec.ts [hunk 38]
```diff
@@ -0,0 +1,89 @@
...
+    connection = settings.connection;
+    connectionSpy = sandbox.spy(connection, "sendDiagnostics");
+  });
+
+  afterEach(() => {
+    connectionSpy.restore();
+  });
+
+  it("Test parsing error validation in xml document", async () => {
+    doc = createTextDocument("xml", ">");
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: doc should be in scope of specific test

File: packages/xml-lsp-server/test/server.spec.ts [hunk 39]
```diff
@@ -0,0 +1,89 @@
...
+} from "vscode-languageserver";
+
+describe("Test XML Language Server", () => {
+  let sandbox: any;
+  let doc: TextDocument;
+  let connection: IConnection;
+  let connectionSpy: any;
+
+  before(() => {
+    sandbox = sinon.createSandbox();
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: maybe the helper should not be aware of `connection` at all.
and only do logic?

File: packages/xml-lsp-server/test/server.spec.ts [hunk 40]
```diff
@@ -0,0 +1,89 @@
...
+        start: pos,
+        end: pos
+      },
+      severity: DiagnosticSeverity.Error,
+      source: SYNTAX_ERROR_MSG
+    };
+
+    const diagnostics: Diagnostic[] = [diagnostic];
+    await validateDocument(connection, doc);
+    assert(connectionSpy.withArgs({ uri: doc.uri, diagnostics }).calledOnce);
```
Comment (bd82) [inline packages/xml-lsp-server/test/server.spec.ts]: we use `chai.expect`

File: packages/xml-lsp-server/tsconfig.json [hunk 41]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "lib": ["es7"],
+    "target": "es2017",
+    "outDir": "out",
+    "rootDir": "src",
+    "sourceMap": true
+  },
+  "include": ["src"],
```
Comment (bd82) [inline packages/xml-lsp-server/tsconfig.json]: I think we can use the strict flag

File: packages/xml-lsp-server/.eslintrc.json (hunk 26 above)
Comment (bd82) [reply packages/xml-lsp-server/.eslintrc.json]: eslint has hierarchy of configuration files, you can re-define the rules only in a test folder

File: packages/xml-lsp-server/package.json [hunk 42]
```diff
@@ -0,0 +1,45 @@
...
+    "node": "10.19.0",
+    "vscode": "1.39.0"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/SAP/xml-tools/"
+  },
+  "dependencies": {
+    "vscode-languageserver": "5.2.1",
+    "@xml-tools/parser": "1.0.2"
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: ensure this is the latest

File: packages/xml-lsp-server/package.json [hunk 43]
```diff
@@ -0,0 +1,45 @@
...
+    "@types/vscode": "1.32.0",
+    "@typescript-eslint/parser": "2.19.0",
+    "@typescript-eslint/eslint-plugin": "2.19.0",
+    "eslint": "6.8.0",
+    "eslint-config-prettier": "6.10.0",
+    "eslint-plugin-prettier": "3.1.2",
+    "mocha": "7.0.1",
+    "nyc": "15.0.0",
+    "prettier": "1.19.1",
+    "sinon": "7.3.2",
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: you probably won't need sinon dep anymore

File: packages/xml-lsp-server/package.json [hunk 44]
```diff
@@ -0,0 +1,45 @@
...
+    "source-map-support": "0.5.16",
+    "ts-node": "8.6.2",
+    "typescript": "3.7.5",
+    "vscode": "^1.1.35",
+    "vscode-test": "1.3.0",
+    "vscode-languageclient": "5.2.1"
+  },
+  "scripts": {
```
Comment (bd82) [inline packages/xml-lsp-server/package.json]: Try to align script names with out packages in this mono-repo

File: packages/xml-lsp-server/src/server.ts [hunk 45]
```diff
@@ -0,0 +1,76 @@
...
+    diagnosticsCache.delete(event.document.uri);
+  }
+});
+
+// Make the document manager listen on the connection
```
Comment (bd82) [inline packages/xml-lsp-server/src/server.ts]: example level comments probably not needed

File: packages/vscode-xml-tools-client/package.json (hunk 11 above)
Comment (bd82) [reply packages/vscode-xml-tools-client/package.json]: update to version 6

**Pull request closed without merging**